openmdao/core/vec_wrapper_complex_step.py

```diff
@@ -0,0 +1,308 @@
+
+        return self.vals[name]
+
+    def __setitem__(self, name, value):
+        """
+        Set the value of the named variable.
+
+        Args
+        ----
+        name : str
+            Name of variable to get the value for.
+
+        value :
+            The unflattened value of the named variable.
+        """
+        self.vals[name] = value
+
+    def __len__(self):
+        """
+        Returns
+        -------
+            The number of keys (variables) in this vector.
+        """
+        return len(self.vecwrap._vardict)
+
+    def __contains__(self, key):
+        """
+        Returns
+        -------
+            A boolean indicating if the given key (variable name) is in this vector.
+        """
+
+        return key in self.vecwrap._vardict
+
+    def __iter__(self):
+        """
+        Returns
+        -------
+            A dictionary iterator over the items in _vardict.
+        """
+        return self.vecwrap._vardict.__iter__()
+
+    def keys(self):
+        """
+        Returns
+        -------
+        list or KeyView (python 3)
+            the keys (variable names) in this vector.
+        """
+        return self.vecwrap._vardict.keys()
+
+    def iterkeys(self):
+        """
+        Returns
+        -------
+        iter of str
+            the keys (variable names) in this vector.
+        """
+        return iterkeys(self.vecwrap._vardict)
+
+    def metadata(self, name):
+        """
+        Returns the metadata for the named variable.
+
+        Args
+        ----
+        name : str
+            Name of variable to get the metadata for.
+
+        Returns
+        -------
+        dict
+            The metadata dict for the named variable.
+
+        Raises
+        -------
+        KeyError
+            If the named variable is not in this vector.
+        """
+        return self.vecwrap._vardict[name]
+
+    def flat(self, name):
+        """
+        Returns flattened value of variable in name.
+
+        Args
+        ----
+        name : str
+            Name of variable to flatten.
+
+        Returns
+        -------
+        ndarray
+            Variable value.
+
+        Raises
+        -------
+        KeyError
+            If the named variable is not in this vector.
+        """
+        val = self.vals[name]
+        if isinstance(val, np.ndarray):
+            return val.flatten()
+        else:
+            return np.array([val])
+
+    def set_complex_var(self, name):
+        """
+        Specifies the current input variable that will be complex stepped.
+
+        Args
+        ----
+        name : str
+            Name of variable to get the metadata for.
+        """
+        var = self.vecwrap.flat[name]
+        self.step_var = name
+        self.step_val = np.zeros(len(var), dtype=np.complex)
+        self.step_val[:] = var
+
+    def unset_complex_var(self):
+        """ Clear out current complex step. """
+        self.step_var = None
+        self.step_val = None
```

Comment:
I can do the first thing.

For the second thing, this is not a user function (and this vector class won't show up in the documentation), so we have control over what goes in, so step_val will never be None.